File: Graph Theory/GraphTheory-main/Lab-13/FindCutVertices.java
//  Prob. 1 Find and print all cut-vertices (articulation points) in an undirected graph. 
import java.util.*;

public class FindCutVertices {
    static void dfs(int u, boolean[] visited, List<List<Integer>> g, int skip) {
        visited[u] = true;
        for (int v : g.get(u)) {
            if (!visited[v] && v != skip) {
                dfs(v, visited, g, skip);
            }
        }
    }

    public static void main(String[] args) {
        int V = 5;
        int[][] edges = {
            {0, 1}, {0, 2}, {2, 3}, {2, 4}
        };

        // Build adjacency list
        List<List<Integer>> g = new ArrayList<>();
        for (int i = 0; i < V; i++) g.add(new ArrayList<>());
        for (int[] e : edges) {
            g.get(e[0]).add(e[1]);
            g.get(e[1]).add(e[0]);
        }

        // Count reachable nodes without removing anything
        boolean[] visited = new boolean[V];
        dfs(0, visited, g, -1);
        int fullCount = 0;
        for (boolean b : visited) if (b) fullCount++;

        System.out.println("Cut Vertices:");
        for (int skip = 0; skip < V; skip++) {
            Arrays.fill(visited, false);

            // pick a start node that's not removed
            int start = (skip == 0) ? 1 : 0;
            dfs(start, visited, g, skip);

            int count = 0;
            for (int i = 0; i < V; i++) {
                if (i != skip && visited[i]) count++;
            }

            if (count < fullCount - 1) {
                System.out.println(skip);
            }
        }
    }
}

File: Graph Theory/GraphTheory-main/FloydAlgorithm/FindCutVertices.c
#include <stdio.h>
#include <stdbool.h>
#include <string.h>

#define V 5   // Number of vertices

// DFS utility function
void dfs(int u, bool visited[], int skip, int graph[V][V]) {
    visited[u] = true;
    for (int v = 0; v < V; v++) {
        if (graph[u][v] && v != skip && !visited[v]) {
            dfs(v, visited, skip, graph);
        }
    }
}

int main() {
    // Graph edges
    int edges[][2] = {
        {0, 1}, {0, 2}, {2, 3}, {2, 4}
    };
    int E = sizeof(edges) / sizeof(edges[0]);

    // Build adjacency matrix
    int graph[V][V] = {0};
    for (int i = 0; i < E; i++) {
        int u = edges[i][0], v = edges[i][1];
        graph[u][v] = 1;
        graph[v][u] = 1;
    }

    // Find reachable nodes without removing anything
    bool visited[V];
    memset(visited, false, sizeof(visited));
    dfs(0, visited, -1, graph);

    int fullCount = 0;
    for (int i = 0; i < V; i++)
        if (visited[i]) fullCount++;

    printf("Cut Vertices:\n");

    // Try removing each vertex
    for (int skip = 0; skip < V; skip++) {
        memset(visited, false, sizeof(visited));

        int start = (skip == 0) ? 1 : 0;
        dfs(start, visited, skip, graph);

        int count = 0;
        for (int i = 0; i < V; i++) {
            if (i != skip && visited[i]) count++;
        }

        if (count < fullCount - 1) {
            printf("%d\n", skip);
        }
    }

    return 0;
}
